File: docs/audits/system-audit/assets/charts.js
(function() {
  var style = getComputedStyle(document.documentElement);
  var accent = style.getPropertyValue('--accent').trim();
  var accent2 = style.getPropertyValue('--accent2').trim();
  var ink = style.getPropertyValue('--ink').trim();
  var muted = style.getPropertyValue('--muted').trim();
  var rule = style.getPropertyValue('--rule').trim();
  var bg2 = style.getPropertyValue('--bg2').trim();
  var bg = style.getPropertyValue('--bg').trim();
  var danger = '#e74c3c';
  var warn = '#f39c12';
  var ok = '#27ae60';

  // --- Chart 1: 问题严重程度分布 ---
  var c1 = echarts.init(document.getElementById('chart-severity'), null, { renderer: 'svg' });
  c1.setOption({
    animation: false,
    tooltip: { appendToBody: true, trigger: 'item', formatter: '{b}: {c}个 ({d}%)' },
    legend: { bottom: 0, textStyle: { color: muted } },
    color: [danger, '#e67e22', warn, '#3498db', ok],
    series: [{
      type: 'pie',
      radius: ['40%', '70%'],
      center: ['50%', '45%'],
      avoidLabelOverlap: false,
      itemStyle: { borderRadius: 6, borderColor: bg, borderWidth: 2 },
      label: { show: true, formatter: '{b}\n{c}个', color: ink, fontSize: 12 },
      labelLine: { lineStyle: { color: rule } },
      data: [
        { value: 1, name: 'P0 阻断' },
        { value: 12, name: 'P1 严重' },
        { value: 18, name: 'P2 中等' },
        { value: 12, name: 'P3 低/质量' },
        { value: 10, name: '优化建议' }
      ]
    }]
  });

  // --- Chart 2: 问题分类分布 ---
  var c2 = echarts.init(document.getElementById('chart-category'), null, { renderer: 'svg' });
  c2.setOption({
    animation: false,
    tooltip: { appendToBody: true, trigger: 'axis', axisPointer: { type: 'shadow' } },
    grid: { left: 120, right: 30, top: 20, bottom: 30 },
    xAxis: { type: 'value', axisLabel: { color: muted }, splitLine: { lineStyle: { color: rule } } },
    yAxis: {
      type: 'category',
      axisLabel: { color: ink, fontSize: 12 },
      axisLine: { lineStyle: { color: rule } },
      data: ['前端缺失/死代码', '事件契约不一致', '架构/理念冲突', '状态机/流程bug', '资源泄漏/并发', 'WebSocket/通信', '安全加固遗留', '代码质量/死代码']
    },
    series: [{
      type: 'bar',
      data: [
        { value: 9, itemStyle: { color: accent } },
        { value: 8, itemStyle: { color: accent2 } },
        { value: 7, itemStyle: { color: '#8e44ad' } },
        { value: 7, itemStyle: { color: danger } },
        { value: 6, itemStyle: { color: warn } },
        { value: 5, itemStyle: { color: '#e67e22' } },
        { value: 5, itemStyle: { color: '#3498db' } },
        { value: 6, itemStyle: { color: ok } }
      ],
      barWidth: 18,
      itemStyle: { borderRadius: [0, 4, 4, 0] },
      label: { show: true, position: 'right', color: ink, fontSize: 12, formatter: '{c}' }
    }]
  });

  // --- Chart 3: 前后端问题分布 ---
  var c3 = echarts.init(document.getElementById('chart-febe'), null, { renderer: 'svg' });
  c3.setOption({
    animation: false,
    tooltip: { appendToBody: true, trigger: 'item' },
    legend: { bottom: 0, textStyle: { color: muted } },
    color: [accent, accent2, '#95a5a6'],
    series: [{
      type: 'pie',
      radius: '65%',
      center: ['50%', '45%'],
      itemStyle: { borderRadius: 4, borderColor: bg, borderWidth: 2 },
      label: { color: ink, formatter: '{b}: {c}个' },
      data: [
        { value: 29, name: '后端' },
        { value: 18, name: '前端' },
        { value: 6, name: '前后端交互' }
      ]
    }]
  });

  // --- Chart 4: 各阶段发现问题数 ---
  var c4 = echarts.init(document.getElementById('chart-area'), null, { renderer: 'svg' });
  c4.setOption({
    animation: false,
    tooltip: { appendToBody: true, trigger: 'axis' },
    legend: { bottom: 0, textStyle: { color: muted } },
    grid: { left: 50, right: 20, top: 40, bottom: 60 },
    xAxis: {
      type: 'category',
      data: ['状态机', '并发/资源', '事件总线', '数据访问', 'WebSocket', '安全/沙箱', '记忆子系统', '前端面板', '前端状态', 'UI/UX'],
      axisLabel: { color: muted, rotate: 30, fontSize: 11 },
      axisLine: { lineStyle: { color: rule } }
    },
    yAxis: { type: 'value', axisLabel: { color: muted }, splitLine: { lineStyle: { color: rule } } },
    series: [
      {
        name: 'P0/P1',
        type: 'bar',
        stack: 'total',
        data: [3, 3, 2, 2, 3, 1, 0, 2, 3, 0],
        itemStyle: { color: danger },
        barWidth: 28
      },
      {
        name: 'P2',
        type: 'bar',
        stack: 'total',
        data: [2, 2, 2, 1, 1, 2, 2, 3, 2, 1],
        itemStyle: { color: warn }
      },
      {
        name: 'P3/优化',
        type: 'bar',
        stack: 'total',
        data: [2, 1, 0, 2, 1, 2, 3, 4, 3, 4],
        itemStyle: { color: ok }
      }
    ]
  });

  window.addEventListener('resize', function() {
    c1.resize(); c2.resize(); c3.resize(); c4.resize();
  });
})();

File: backend/eval_v2/reports/charts.py
"""ECharts configuration generators for HTML report visualizations."""

from __future__ import annotations

import json
from typing import Any

BRAND_COLOR = "#335c8e"
ACCENT_COLORS = ["#335c8e", "#5b8db8", "#e07a5f", "#81b29a", "#f2cc8f", "#9d8189", "#6c757d", "#adb5bd"]


def _json(data: Any) -> str:
    return json.dumps(data, ensure_ascii=False, default=str)


def overview_cards(result: Any) -> str:
    """Generate overview metric cards HTML."""
    ci = result.pass_at_1_ci
    cards = [
        ("Pass@1", f"{result.pass_at_1:.1%}", f"95% CI: [{ci[0]:.1%}, {ci[1]:.1%}]"),
        ("Pass@3", f"{result.pass_at_3:.1%}", f"Stable: {result.pass_at_3_stable:.1%}"),
        ("Avg Score", f"{result.avg_aggregate_score:.1f}", "/100"),
        ("Total Cost", f"${result.total_cost_usd:.3f}", f"${result.cost_per_pass:.3f}/pass"),
        ("Cases", str(result.n_cases), f"{result.n_runs_per_case} runs each"),
        (
            "Duration",
            f"{result.duration_seconds:.0f}s",
            f"p50={result.latency_p50_ms / 1000:.1f}s, p95={result.latency_p95_ms / 1000:.1f}s",
        ),
    ]
    html = '<div class="cards-grid">'
    for title, value, subtitle in cards:
        html += f"""
        <div class="card">
            <div class="card-title">{title}</div>
            <div class="card-value">{value}</div>
            <div class="card-subtitle">{subtitle}</div>
        </div>"""
    html += "</div>"
    return html


def layer_pass_chart(result: Any) -> tuple[str, str]:
    """Layer pass rates bar chart."""
    layers = []
    rates = []
    labels = {
        "l0_health": "L0 Health",
        "l1_process": "L1 Process",
        "l2_internal": "L2 Internal",
        "l3_semantic": "L3 Semantic",
        "l4_comparative": "L4 Comparative",
    }
    for k, v in result.layer_pass_rates.items():
        if k in labels:
            layers.append(labels.get(k, k))
            rates.append(round(v * 100, 1))

    chart_id = "chart-layer-pass"
    option = {
        "tooltip": {"trigger": "axis"},
        "xAxis": {"type": "category", "data": layers, "axisLabel": {"rotate": 15}},
        "yAxis": {"type": "value", "max": 100, "axisLabel": {"formatter": "{value}%"}},
        "series": [
            {
                "type": "bar",
                "data": [{"value": r, "itemStyle": {"color": BRAND_COLOR if r >= 60 else "#e07a5f"}} for r in rates],
                "label": {"show": True, "position": "top", "formatter": "{c}%"},
            }
        ],
        "grid": {"left": 50, "right": 20, "bottom": 50, "top": 30},
    }
    return chart_id, _json(option)


def dimension_radar(result: Any) -> tuple[str, str]:
    """Dimension score radar chart."""
    dims = list(result.dimension_scores.keys())
    scores = [round(result.dimension_scores.get(d, 0) * 100, 1) for d in dims]

    chart_id = "chart-dim-radar"
    if not dims:
        option = {"title": {"text": "No dimension data", "left": "center"}}
    else:
        option = {
            "tooltip": {},
            "radar": {
                "indicator": [{"name": d, "max": 100} for d in dims],
                "shape": "polygon",
                "splitNumber": 4,
            },
            "series": [
                {
                    "type": "radar",
                    "data": [
                        {
                            "value": scores,
                            "name": "Score",
                            "areaStyle": {"color": "rgba(51,92,142,0.2)"},
                            "lineStyle": {"color": BRAND_COLOR},
                            "itemStyle": {"color": BRAND_COLOR},
                        }
                    ],
                }
            ],
        }
    return chart_id, _json(option)


def failure_pie(result: Any) -> tuple[str, str]:
    """Failure mode pie chart."""
    data = []
    label_map = {
        "infrastructure": "Infrastructure",
        "model": "Model/System",
        "test_bug": "Test Bug",
        "none": "Passed",
    }
    for k, v in result.failure_mode_breakdown.items():
        if v > 0:
            data.append({"name": label_map.get(k, k), "value": v})

    chart_id = "chart-failure-pie"
    color_map = {"Passed": "#81b29a", "Infrastructure": "#e07a5f", "Model/System": "#f2cc8f", "Test Bug": "#9d8189"}
    option = {
        "tooltip": {"trigger": "item", "formatter": "{b}: {c} ({d}%)"},
        "series": [
            {
                "type": "pie",
                "radius": ["40%", "70%"],
                "data": [
                    {
                        "value": d["value"],
                        "name": d["name"],
                        "itemStyle": {"color": color_map.get(d["name"], "#6c757d")},
                    }
                    for d in data
                ],
                "label": {"formatter": "{b}: {c}"},
            }
        ],
    }
    return chart_id, _json(option)


def cost_pareto(result: Any) -> tuple[str, str]:
    """Cost vs Quality Pareto scatter plot."""
    points = []
    for r in result.per_case_results:
        if r.run_index == 0:
            points.append([round(r.cost.total_cost_usd, 4), round(r.aggregate_score, 1), r.case_id, r.passed])

    chart_id = "chart-pareto"
    option = {
        "tooltip": {
            "trigger": "item",
            "formatter": "function(p) { return p.data[2] + '<br/>Cost: $' + p.data[0] + '<br/>Score: ' + p.data[1]; }",
        },
        "xAxis": {"name": "Cost (USD)", "type": "value"},
        "yAxis": {"name": "Score", "type": "value", "min": 0, "max": 100},
        "series": [
            {
                "type": "scatter",
                "symbolSize": 12,
                "data": points,
                "itemStyle": {
                    "color": "function(params) { return params.data[3] ? '#81b29a' : '#e07a5f'; }",
                },
            }
        ],
        "grid": {"left": 60, "right": 20, "bottom": 50, "top": 30},
    }
    return chart_id, _json(option)


def ablation_bar(result: Any) -> tuple[str, str]:
    """Ablation comparison bar chart."""
    chart_id = "chart-ablation"
    if not result.ablation:
        option = {"title": {"text": "No ablation data", "left": "center"}}
    else:
        ab = result.ablation
        option = {
            "tooltip": {"trigger": "axis"},
            "xAxis": {"type": "category", "data": ["Multi-Agent", "Direct LLM"]},
            "yAxis": {"type": "value", "max": 100, "name": "Score"},
            "series": [
                {
                    "type": "bar",
                    "data": [round(ab.multi_agent_score * 100, 1), round(ab.direct_llm_score * 100, 1)],
                    "itemStyle": {"color": BRAND_COLOR},
                    "label": {"show": True, "position": "top"},
                },
            ],
            "grid": {"left": 50, "right": 20, "bottom": 30, "top": 30},
        }
    return chart_id, _json(option)


def dimension_bar(result: Any) -> tuple[str, str]:
    """Dimension score bar chart (grouped by type)."""
    dims = list(result.dimension_scores.keys())
    scores = [round(result.dimension_scores.get(d, 0) * 100, 1) for d in dims]

    chart_id = "chart-dim-bar"
    option = {
        "tooltip": {"trigger": "axis"},
        "xAxis": {"type": "category", "data": dims, "axisLabel": {"rotate": 30, "fontSize": 10}},
        "yAxis": {"type": "value", "max": 100, "name": "Score"},
        "series": [
            {
                "type": "bar",
                "data": [{"value": s, "itemStyle": {"color": BRAND_COLOR if s >= 60 else "#e07a5f"}} for s in scores],
                "label": {"show": True, "position": "top", "fontSize": 9},
            }
        ],
        "grid": {"left": 50, "right": 20, "bottom": 80, "top": 30},
    }
    return chart_id, _json(option)


def cost_stacked(result: Any) -> tuple[str, str]:
    """Cost breakdown stacked bar."""
    chart_id = "chart-cost-stacked"
    per_case_sut = []
    per_case_judge = []
    case_ids = []
    for r in result.per_case_results:
        if r.run_index == 0:
            case_ids.append(r.case_id[:20])
            per_case_sut.append(round(r.cost.sut_cost_usd, 4))
            per_case_judge.append(round(r.cost.judge_cost_usd, 4))

    option = {
        "tooltip": {"trigger": "axis"},
        "xAxis": {"type": "category", "data": case_ids, "axisLabel": {"rotate": 45, "fontSize": 8}},
        "yAxis": {"type": "value", "name": "Cost (USD)"},
        "series": [
            {"name": "SUT", "type": "bar", "stack": "cost", "data": per_case_sut, "itemStyle": {"color": BRAND_COLOR}},
            {
                "name": "Judge",
                "type": "bar",
                "stack": "cost",
                "data": per_case_judge,
                "itemStyle": {"color": "#5b8db8"},
            },
        ],
        "legend": {"top": 0},
        "grid": {"left": 60, "right": 20, "bottom": 100, "top": 40},
    }
    return chart_id, _json(option)
